src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,134 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+
+    private static final int INITIAL_CAPACITY = 10;
+
+    private Object[] objects;
+    private int size;
+
+    public ArrayList() {
+        this.objects = new Object[INITIAL_CAPACITY];
+    }
+
     @Override
     public void add(T value) {
-
+        if (size() == objects.length) {
+            grow();
+        }
+        objects[size()] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index > size() || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index is invalid!");
+        }
+        if (index == size() - 1 && index == size()) {
+            this.add(value);
+        } else {
+            shift(index);
+            objects[index] = value;
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        if (!list.isEmpty()) {
+            for (int i = 0; i < list.size(); i++) {
+                this.add(list.get(i));
+            }
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (checkIndex(index)) {
+            return (T) objects[index];
+        }
+        throw new ArrayListIndexOutOfBoundsException("Index is invalid!");
     }
 
     @Override
     public void set(T value, int index) {
-
+        if (checkIndex(index)) {
+            objects[index] = value;
+        } else {
+            throw new ArrayListIndexOutOfBoundsException("Index is invalid!");
+        }
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (checkIndex(index)) {
+            Object element = objects[index];
+            collapse(index);
+            size--;
+            return (T) element;
+        }
+        throw new ArrayListIndexOutOfBoundsException("Index is invalid!");
```

Comment:
```suggestion
        checkIndex(index)
        T removedElement = (T) objects[index];
        System.arraycopy(objects, index + 1, objects, index, size - index - 1);
        size--;
        return removedElement;
```
simply use 1 array copy

Reply:
OK